### ISSUE TITLE
Analyzer benchmarks regression

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#31419]
I don't understand why or how, but https://github.com/flutter/flutter/pull/12977 managed to cause the Analyzer benchmarks to regress across the board, especially the flutter_repo benchmarks.

![screen shot 2017-11-20 at 9 45 55 pm](https://user-images.githubusercontent.com/551196/33056741-333a5b84-ce3c-11e7-91ff-73232256c1a9.png)

cc @pq @devoncarew

**Comment by Hixie** [comment]
(See also https://github.com/dart-lang/sdk/issues/31301 which is about a longer-term trend in these benchmarks.)

**Comment by pq** [comment]
cc @bwilkerson @scheglov

**Comment by scheglov** [comment]
Honestly, I don't understand what I'm looking at.
Do you analyze some code base, and it now takes longer?
Or it is not just analyze, but some scenarios like analyze - edit - analyze?

**Comment by pq** [comment]
IIRC this is time to analyze the flutter repo codebase.

**Comment by scheglov** [comment]
Hm... `64.0 ms` does not look as time I would expect is enough to analyze the flutter repo.

**Comment by pq** [comment]
Indeed, `64.0 ms` would be awesome!  Unfortunately, we're looking at **seconds**. ☹️ 

That said, looking at how analysis is happening there are at least a few things to explore.

1. the code path in `flutter analyze` that analyzes the repo needs updating to the new driver APIs.  (While this may not provide speeding up, it needs to happen regardless and may well give us a boost.)
2. regarding slowness in `watch`, which takes another path, looking at go/lem confirms a gradual slow down of analysis over time.  To understand this better, we'd need to do some profiling.
3. in general, are we doing any caching?  Could summaries help here?  Are all of these benchmarks on a cold start?

**Comment by Hixie** [comment]
There are all cold-start, yes.

I'll update the units to seconds, not sure why I never noticed that!

**Comment by devoncarew** [comment]
At least for item 1 above, that would hopefully be covered by a general conversion to just using the analysis server (and not a custom client of the package:analyzer APIs).

**Comment by pq** [comment]
> At least for item 1 above, that would hopefully be covered by a general conversion to just using the analysis server (and not a custom client of the package:analyzer APIs).

Totally.  Unfortunately, #31301 suggests that the issue is in `watch` (if the benchmark name `flutter_repo_watch` is to be trusted) which means item 2 is the long-term issue to track.

As for 1, comments in the [analyze_once source](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/analyze_once.dart#L55) mention lack of a dartdoc summary as the reason we're still using APIs directly but maybe this is something we should revisit?

cc @danrubel

**Comment by devoncarew** [comment]
Yup, just pointing out that moving to using the analysis server will solve a few issues :)

I think we're at a point where we can enable that lint by default, and fail builds if there are any violations (i.e., we no longer need to count undocumented members). Should allow us to move forward with --watch as the only implementation of flutter analyze.

We will need to roll to a new version of the lint, one with the the test/ dir exclusions.

**Comment by pq** [comment]
That would be ideal.  If I'm reading things right, that fix made it into the linter v `0.1.38` (https://github.com/dart-lang/linter/commit/ed469a2f0f9399f7b9ba3a824deb1b7fb33481dd), which has been in the SDK since https://github.com/dart-lang/sdk/commit/87820ba6866aa2c6d5d5eabcaceaa7ee5420adcd so we'd need to bump `dart-sdk.version` to at least [2.0.0-dev.6.0](https://github.com/dart-lang/sdk/blob/2.0.0-dev.6.0/DEPS#L94).

**Comment by Hixie** [comment]
We definitely still use the `--dartdocs` feature. I use it frequently, to track down which members need documenting, and we have a benchmark that uses it to track the number of undocumented members. We also have that metric in our OKRs.

We do not want to fail builds that have undocumented members (and never will), because it encourages people to write poor documentation. I would rather have no documentation checked in for a member than have poor documentation, because I can't track down poor documentation.

**Comment by Hixie** [comment]
These benchmarks have continued to regress and we're now entirely red for `flutter_repo_batch` (at almost 40s, it used to be just over 30s), frequently red for `flutter_repo_watch` (at around 65s, it used to be 50s), and consistently pale green for mega_gallery_batch (25s, was 22s) and mega_gallery_watch (24s, was 20s).

**Comment by Hixie** [comment]
Since it's been so long, the specific regression that originally triggered me to file this bug is probably buried in the noise among other regressions, so I'm going to merge this and https://github.com/dart-lang/sdk/issues/31301 by closing that other one.
